### PR TITLE
Feat: Added support to mysql on some raw queries

### PR DESCRIPTION
**Pull request by darthaud**
Signed-off-by: Daniel Dias Branco Arthaud <arthaud@gmail.com>

## Hey, I just made a Pull Request!

I tweaked some queries to support mysql.

issue associated [https://github.com/backstage/backstage/issues/6281](https://github.com/backstage/backstage/issues/6281)

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend-tasks/src/tasks/TaskWorker.ts [hunk 1]
```diff
@@ -292,14 +292,23 @@ export class TaskWorker {
...
+          `max(datetime(next_run_start_at, ?), datetime('now'))`,
+          [`+${dt} seconds`],
+        );
+      }
+
+      if (this.knex.client.config.client.includes('mysql')) {
+        nextRun = this.knex.raw(
+          `greatest(next_run_start_at + interval ${dt} second, now())`,
+        );
+      }
```
Comment (freben) [inline packages/backend-tasks/src/tasks/TaskWorker.ts:311]: should change this into `if (sqlite) {...} else if (mysql) {...} else {...}`, so we don't needlessly construct a fragment on like 296 that then gets thrown away.

File: plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts [hunk 2]
```diff
@@ -113,7 +113,10 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
 
     // Delete old relations
     let previousRelationRows: DbRelationsRow[];
-    if (tx.client.config.client.includes('sqlite3')) {
+    if (
+      tx.client.config.client.includes('sqlite3') ||
+      tx.client.config.client.includes('mysql')
```
Comment (freben) [inline plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:118]: oh really? i would have thought that the mysql driver supported RETURNING just fine?

**Review by freben — COMMENTED** [review]
Thanks. Nice that you do this in digestible pieces.

Now, we should perhaps worry about this degrading over time, unless it's backed by tests, right?

File: plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts (hunk 2 above)
Comment (darthaud) [reply plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:118]: Unfortunately. When I put mysql as client this snippet throws the following warning (.returning() is not supported by mysql and will have no effect) and in another snippet it throws the following error (failed previous.relations.map is not a function type =plugin stack=TypeError: previous.relations.map is not a function)

Comment (darthaud) [reply plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:118]: If you don't mind. I think that we could use the same solution, without returning, to all databases...

Comment (darthaud) [reply plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:118]: @freben What do you think about it?

**Comment by darthaud** [comment]
> Thanks. Nice that you do this in digestible pieces.
> 
> Now, we should perhaps worry about this degrading over time, unless it's backed by tests, right?

@freben 

I created tests for nowPlus util. And I try to enable mysql on the other tests, but we need to fix the migrations too.

What do you think is the best approach to do?

**Comment by freben** [comment]
Unfortunately, once you start going after the migrations, that's when the whole thing starts to get complex. It'll quickly become a much larger and more complex pull request. So maybe we can do such things separately later on.

Regarding skipping `returning` - that would be extremely unfortunate since it'll in some cases be necessary to be able to have atomicity guarantees. So we'd prefer continuing to use that whenever possible.

**Comment by darthaud** [comment]
> Unfortunately, once you start going after the migrations, that's when the whole thing starts to get complex. It'll quickly become a much larger and more complex pull request. So maybe we can do such things separately later on.
> 
> Regarding skipping `returning` - that would be extremely unfortunate since it'll in some cases be necessary to be able to have atomicity guarantees. So we'd prefer continuing to use that whenever possible.

Ok! I will keep the returning.

About the tests, I'm going to remove Mysql8 from the databases so it doesn't break anymore. Is there anything else we can do about the tests?

**Comment by freben** [comment]
Probably not at this point 🤔 Since the migrations are a blocker.

**Comment by darthaud** [comment]
> Probably not at this point thinking Since the migrations are a blocker.

Maybe we can talk about the migrations on the issue. Because we have some issues to solve.

**Pull request merged**